nb_anacondacloud/static/main.js

```diff
@@ -1,73 +1,87 @@
+        '/latest/img/AnacondaCloud_logo_green.png';
 
 
-        if (attachEnv) {
-            envName = Jupyter.notebook.kernel.name;
-        } else {
-            envName = null;
+    function api(route){
+      // reusable API mapping, handles
+      return Jupyter.notebook.base_url + NS + '/' + route;
+    }
+
+
+    function metadata(key, value){
+        // wrapper for metadata, ensuring the top-level namespace
+        var md = Jupyter.notebook.metadata;
+
+        // ensure top-level namespace is initialized
+        md[NS] = md[NS] || {};
+
+        if (!arguments.length) {
+          return md[NS];
+        } else if(arguments.length === 1){
+          return md[NS][key];
+        } else{
+          md[NS][key] = value;
         }
+    }
+
+
+    function publishNotebook(){
+        metadata('environment', metadata('attach-environment') ?
+            Jupyter.notebook.kernel.name :
+            null);
+        Jupyter.notebook.save_notebook().then(uploadNotebook);
+    }
+
+
+    function uploadNotebook() {
+        var interval;
 
         $.ajax({
-            url: "/ac-publish",
-            method: 'POST',
-            dataType: 'json',
-            contentType: 'application/json; charset=utf-8',
-            processData: false,
-            data: JSON.stringify({
-                name: notebookName,
-                organization: anacondaOrganization,
-                content: nbj,
-                envName: envName
-            })
-        }).done(function(data) {
-            Jupyter.notification_area.get_widget("notebook").
-                set_message("Your notebook has been uploaded.", 4000);
-            Jupyter.notebook.metadata.anacondaCloudURL = data.url;
-            Jupyter.notebook.save_notebook();
-            updateVisitLink(data.url);
-        }).fail(function(jqXHR, textStatus) {
-            var notif, title, body;
-            if (jqXHR.status == 401) {
-                showUnauthorized();
-            } else {
-                notif = 'Error: ' + jqXHR.statusText;
-            }
-            IPython.notification_area.get_widget("notebook").
-                danger(notif, 4000);
-        }).always(function(data, textStatus) {
-            clearInterval(interval);
-        });
+              url: api('publish'),
+              method: 'POST',
+              dataType: 'json',
+              contentType: 'application/json; charset=utf-8',
+              processData: false,
+              data: JSON.stringify({
+                  name: Jupyter.notebook.notebook_name,
+                  content: Jupyter.notebook.toJSON()
+              })
+          })
+          .done(function(data) {
+              Jupyter.notification_area.get_widget('notebook')
+                  .set_message(Jupyter.notebook.notebook_name +
+                      ' was published',
+                      4000);
+              metadata('url', data.url);
+              Jupyter.notebook.save_notebook();
```

Comment:
the metadata will have changed.

What would happen otherwise is:
- user writes notebook
- user fills out details, publishes
  - _triggers automatic save_
  - _call to REST_
  - _metadata is updated with new url, etc_
  - ~~_triggers automatic save_~~
- user navigates away without saving (browser crash, or get pissed because publish is hanging everything)
  - **_the record of publishing is discarded_**
- user returns later
- **user sees no indication that they had ever published**
- **user files bug**